File: idioms/javascript/SharedContextClosures.js
#! /usr/bin/env node
"use strict";

/**
 * Create two closures with share context
 * @param value - Value of context
 */
function CreateClosures(value) {
  let context = JSON.parse(
    JSON.stringify(value)
  );
  function GetMethod() {
    return JSON.parse(
      JSON.stringify(context)
    );
  }
  function SetMethod(value) {
    context = JSON.parse(
      JSON.stringify(value)
    );
  }
  return {
    'getContext': GetMethod,
    'setContext': SetMethod
  };
}

// Creating closures with protective share context
// Object mutability is irrelevant because context is
// protected by the closure (capture by copy and unreachable)
const closures = CreateClosures(['A']);

// Print original context value
console.log('get context value: ', closures.getContext());

// Set new context value
closures.setContext(['B']);
console.log('set and get context value: ', closures.getContext());

// It is not possible to get a refence to context
const context = closures.getContext();
context[0] = 'C';
console.log('get context value: ', closures.getContext());

File: idioms/python/SharedContextClosures.py
#! /usr/bin/env python
import copy

def CreateClosure(value):
    """Create two closures with share context."""
    context = copy.deepcopy(value)
    def GetContext():
        return copy.deepcopy(context)
    def SetContext(value):
        nonlocal context
        context = copy.deepcopy(value)
    return GetContext, SetContext

# Creating closures with protective share context
# Object mutability is irrelevant because context is
# protected by the closure (capture by copy and unreachable)
get_context, set_context = CreateClosure(['A'])

# Print original context value
print('get context value: ', get_context())

# Set new context value
set_context(['B'])
print('set and get context value: ', get_context())

# It is not possible to get a refence to context
context = get_context()
context[0] = 'C'
print('get context value: ', get_context())
